// dicom/lib/pdu.js
// Generated by CoffeeScript 1.8.0
(function() {
  var AbstractSyntaxItem, ApplicationContextItem, AsynchronousOperationsWindowItem, ITEM_BY_NAME, ITEM_BY_TYPE, ImplementationClassUidItem, ImplementationVersionNameItem, Item, MaximumLengthItem, PDU, PDUAssociateAc, PDUAssociateRq, PDUDecoder, PDUEncoder, PDU_BY_NAME, PDU_BY_TYPE, PresentationContextItem, PresentationContextItemAc, ScpScuRoleSelectionItem, TransferSyntaxItem, UserInformationItem, ZERO_BUFF, echo_scu, item_constructor, log, mk_uint16, mk_uint32, net, pdu_by_name, pdu_constructor, printf, readbuffer, stream, total_length, trim_ui, vrs,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  stream = require("stream");

  printf = require("printf");

  readbuffer = require("./readbuffer");

  vrs = require("../lib/vrs");

  log = require("./logger")('pdu');

  total_length = function(arr) {
    return arr.reduce(function(sum, buff) {
      return sum + buff.length;
    }, 0);
  };

  PDUDecoder = (function(_super) {
    __extends(PDUDecoder, _super);

    function PDUDecoder(options) {
      if (!(this instanceof PDUDecoder)) {
        return new PDUDecoder(options);
      }
      PDUDecoder.__super__.constructor.call(this, options);
      this._writableState.objectMode = false;
      this._readableState.objectMode = true;
      this.__buffer = readbuffer();
      this.__saved = this.__buffer.copy();
    }

    PDUDecoder.prototype._transform = function(chunk, encoding, cb) {
      this.__buffer.push(chunk);
      if (log.debug()) {
        log.debug({
          buffer: this.__buffer.log_summary()
        }, "_transform");
      }
      this.__consume_pdu();
      if (log.debug()) {
        log.debug({
          buffer: this.__buffer.log_summary()
        }, "_transform done, calling cb");
      }
      return cb();
    };

    PDUDecoder.prototype._flush = function(cb) {
      log.debug("_flush");
      return cb();
    };

    PDUDecoder.prototype.__consume_pdu = function() {
      var err, __constr, __header, __length, __pdu, __pdubuff, __type;
      try {
        this.__saved = this.__buffer.copy();
        __header = this.__buffer.consume(6);
        __type = __header[0];
        __length = __header.readUInt32BE(2);
        __pdubuff = this.__buffer.consume(__length);
        __constr = pdu_constructor(__type);
        if (__constr == null) {
          throw new vrs.DicomError("Unrecognized PDU: " + __type);
        }
        __pdu = new __constr(__pdubuff);
        log.trace({
          pdu: __pdu
        }, "__consume_pdu");
        return this.push(__pdu);
      } catch (_error) {
        err = _error;
        if (err != null ? err.needMoreInput : void 0) {
          this.__buffer = this.__saved;
          return log.debug({
            needMoreInput: err.needMoreInput,
            buffer: this.__buffer.log_summary(),
            error: err
          }, "_action_wrapper: restored buffer after NeedMoreInput");
        } else {
          log.error({
            error: err
          }, "__consume_pdu: error");
          return this.emit('error', err);
        }
      }
    };

    return PDUDecoder;

  })(stream.Transform);

  PDU = (function() {
    PDU.prototype._json_name = true;

    PDU.prototype._single_value = false;

    function PDU(buff_or_json) {
      if (buff_or_json instanceof Buffer) {
        this._buff = buff_or_json;
        this.decode();
        delete this._buff;
      } else {
        this.from_json(buff_or_json);
      }
      return this;
    }

    PDU.prototype.decode_var_items = function(start, end) {
      var _cnt, _item, _name;
      if (log.trace()) {
        log.trace({
          start: start,
          end: end
        }, "PDU.decode_var_items");
      }
      while (start < end) {
        if (log.trace()) {
          log.trace({
            start: start,
            end: end
          }, "PDU.decode_var_items");
        }
        _item = this.decode_item(start);
        _name = _item.name;
        _cnt = this.var_item_counts[_name];
        if (_cnt === 1) {
          if (this[_name]) {
            throw new vrs.DicomError("Only one " + _name + " allowed");
          } else {
            this[_name] = _item;
          }
        } else {
          if (this[_name] == null) {
            this[_name] = [];
          }
          this[_name].push(_item);
        }
        start = _item._end;
      }
      return void 0;
    };

    PDU.prototype.decode_item = function(start) {
      var _constr, _item, _length, _type;
      _type = this._buff[start];
      _length = this._buff.readUInt16BE(start + 2);
      _constr = item_constructor(_type);
      if (_constr == null) {
        log.warn({
          type: printf("%02X", _type),
          start: start,
          length: _length
        }, "PDU item not implemented");
        return {
          type: _type,
          name: 'unknown',
          _start: start,
          _end: start + 4 + _length
        };
      } else {
        _item = new _constr(this._buff, start, start + 4 + _length);
        if (log.trace()) {
          log.trace(_item.log_summary(), "decoded item");
        }
        return _item;
      }
    };

    PDU.prototype.log_summary = function() {
      var k, v, _summary;
      _summary = {};
      for (k in this) {
        v = this[k];
        if (k !== '_buff') {
          if ((v != null ? v.log_summary : void 0) != null) {
            v = v.log_summary();
          }
          _summary[k] = v;
        }
      }
      return _summary;
    };

    PDU.prototype.to_json = function() {
      var _item, _item_value, _json, _k, _ref, _v;
      if (this.value != null) {
        return this.value;
      }
      _json = this._json_name ? {
        name: this.name
      } : {};
      _item_value = function(item) {
        if ((item != null ? item.to_json : void 0) != null) {
          item = item.to_json();
        }
        return item;
      };
      _ref = this.var_item_counts;
      for (_k in _ref) {
        _v = _ref[_k];
        if (_v === 1) {
          _item = this[_k];
          _json[_k] = _item_value(_item);
        } else if (this[_k] != null) {
          _json[_k] = (function() {
            var _i, _len, _ref1, _results;
            _ref1 = this[_k];
            _results = [];
            for (_i = 0, _len = _ref1.length; _i < _len; _i++) {
              _item = _ref1[_i];
              _results.push(_item_value(_item));
            }
            return _results;
          }).call(this);
        }
      }
      return _json;
    };

    PDU.prototype.from_json = function(json) {
      var _constr, _k, _ref, _v, _x;
      if (this._single_value) {
        this.value = json;
        return;
      }
      _ref = this.var_item_counts;
      for (_k in _ref) {
        _v = _ref[_k];
        _constr = ITEM_BY_NAME[_k];
        if (_constr == null) {
          throw new vrs.DicomError("no such item: " + _k);
        }
        log.trace({
          name: _k,
          count: _v,
          constr: _constr
        }, "named item");
        if (_v === 1) {
          this[_k] = new _constr(json[_k]);
        } else {
          if (json[_k]) {
            this[_k] = (function() {
              var _i, _len, _ref1, _results;
              _ref1 = json[_k];
              _results = [];
              for (_i = 0, _len = _ref1.length; _i < _len; _i++) {
                _x = _ref1[_i];
                _results.push(new _constr(_x));
              }
              return _results;
            })();
          }
        }
        log.trace({
          json: json[_k],
          result: this[_k]
        }, "from json result");
      }
    };

    return PDU;

  })();

  PDUAssociateRq = (function(_super) {
    __extends(PDUAssociateRq, _super);

    function PDUAssociateRq() {
      return PDUAssociateRq.__super__.constructor.apply(this, arguments);
    }

    PDUAssociateRq.prototype.type = 0x01;

    PDUAssociateRq.prototype.name = 'association_rq';

    PDUAssociateRq.prototype.var_item_counts = {
      application_context: 1,
      presentation_context: -1,
      user_information: 1
    };

    PDUAssociateRq.prototype.decode = function() {
      this._protocol_version = this._buff.readUInt16BE(0);
      this.called_aet_title = this._buff.slice(4, 20).toString().trim();
      this.calling_aet_title = this._buff.slice(20, 36).toString().trim();
      return this.decode_var_items(68, this._buff.length);
    };

    PDUAssociateRq.prototype.encode = function() {
      var item, _buffers, _header, _i, _len, _ref, _var_len;
      _buffers = [''];
      _buffers.push(this.application_context.encode());
      _ref = this.presentation_context;
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        item = _ref[_i];
        _buffers.push(item.encode());
      }
      _buffers.push(this.user_information.encode());
      _var_len = total_length(_buffers);
      _header = Buffer.concat([new Buffer([0x01, 0x00]), mk_uint32(68 + _var_len), mk_uint16(1), new Buffer([0x00, 0x00]), new Buffer(printf("%-16s%-16s", this.called_aet_title.substr(0, 16), this.calling_aet_title.substr(0, 16)), 'binary'), ZERO_BUFF.slice(0, 32)]);
      _buffers[0] = _header;
      return Buffer.concat(_buffers);
    };

    PDUAssociateRq.prototype.to_json = function() {
      var _json;
      _json = PDUAssociateRq.__super__.to_json.call(this);
      _json.called_aet_title = this.called_aet_title;
      _json.calling_aet_title = this.calling_aet_title;
      return _json;
    };

    PDUAssociateRq.prototype.from_json = function(json) {
      this.called_aet_title = json.called_aet_title;
      this.calling_aet_title = json.calling_aet_title;
      return PDUAssociateRq.__super__.from_json.call(this, json);
    };

    return PDUAssociateRq;

  })(PDU);

  PDUAssociateAc = (function(_super) {
    __extends(PDUAssociateAc, _super);

    function PDUAssociateAc() {
      return PDUAssociateAc.__super__.constructor.apply(this, arguments);
    }

    return PDUAssociateAc;

  })(PDUAssociateRq);

  Item = (function(_super) {
    __extends(Item, _super);

    Item.prototype._json_name = false;

    function Item(buff_or_json, start, end) {
      if (buff_or_json instanceof Buffer) {
        this._start = start;
        this._end = end;
      }
      Item.__super__.constructor.call(this, buff_or_json);
    }

    Item.prototype.ui_str = function(offset) {
      var _start, _str, _ui;
      _start = this._start + offset;
      _str = this._buff.toString('binary', _start, this._end);
      _ui = trim_ui(_str);
      return _ui;
    };

    Item.prototype.encode_value_str = function() {
      return Buffer.concat([new Buffer([this.type, 0]), mk_uint16(this.value.length), new Buffer(this.value, 'binary')]);
    };

    return Item;

  })(PDU);

  ApplicationContextItem = (function(_super) {
    __extends(ApplicationContextItem, _super);

    function ApplicationContextItem() {
      return ApplicationContextItem.__super__.constructor.apply(this, arguments);
    }

    ApplicationContextItem.prototype.type = 0x10;

    ApplicationContextItem.prototype.name = 'application_context';

    ApplicationContextItem.prototype._single_value = true;

    ApplicationContextItem.prototype.decode = function() {
      return this.value = this.ui_str(4);
    };

    ApplicationContextItem.prototype.encode = function() {
      return this.encode_value_str();
    };

    return ApplicationContextItem;

  })(Item);

  PresentationContextItem = (function(_super) {
    __extends(PresentationContextItem, _super);

    function PresentationContextItem() {
      return PresentationContextItem.__super__.constructor.apply(this, arguments);
    }

    PresentationContextItem.prototype.type = 0x20;

    PresentationContextItem.prototype.name = 'presentation_context';

    PresentationContextItem.prototype.var_item_counts = {
      abstract_syntax: 1,
      transfer_syntax: -1
    };

    PresentationContextItem.prototype.decode = function() {
      this.id = this._buff[this._start + 4];
      return this.decode_var_items(this._start + 8, this._end);
    };

    PresentationContextItem.prototype.encode = function() {
      var _buffers, _header, _len, _ts;
      _buffers = [
        '', this.fixed_fields(), this.abstract_syntax.encode(), Buffer.concat((function() {
          var _i, _len, _ref, _results;
          _ref = this.transfer_syntax;
          _results = [];
          for (_i = 0, _len = _ref.length; _i < _len; _i++) {
            _ts = _ref[_i];
            _results.push(_ts.encode());
          }
          return _results;
        }).call(this))
      ];
      _len = total_length(_buffers);
      _header = Buffer.concat([new Buffer([0x20, 0]), mk_uint16(_len)]);
      _buffers[0] = _header;
      return Buffer.concat(_buffers);
    };

    PresentationContextItem.prototype.fixed_fields = function() {
      return new Buffer([this.id, 0, 0, 0]);
    };

    PresentationContextItem.prototype.to_json = function() {
      var _json;
      _json = PresentationContextItem.__super__.to_json.call(this);
      _json.id = this.id;
      return _json;
    };

    PresentationContextItem.prototype.from_json = function(json) {
      this.id = json.id;
      return PresentationContextItem.__super__.from_json.call(this, json);
    };

    return PresentationContextItem;

  })(Item);

  PresentationContextItemAc = (function(_super) {
    __extends(PresentationContextItemAc, _super);

    function PresentationContextItemAc() {
      return PresentationContextItemAc.__super__.constructor.apply(this, arguments);
    }

    PresentationContextItemAc.prototype.type = 0x21;

    PresentationContextItemAc.prototype.name = 'presentation_context_ac';

    PresentationContextItemAc.prototype.decode = function() {
      PresentationContextItemAc.__super__.decode.call(this);
      return this.resultReason = this._buff[this._start + 6];
    };

    PresentationContextItemAc.prototype.fixed_fields = function() {
      return new Buffer([this.id, 0, this.resultReason, 0]);
    };

    PresentationContextItemAc.prototype.to_json = function() {
      var _json;
      _json = PresentationContextItemAc.__super__.to_json.call(this);
      _json.resultReason = this.resultReason;
      return _json;
    };

    PresentationContextItemAc.prototype.from_json = function(json) {
      this.resultReason = json.resultReason;
      return PresentationContextItemAc.__super__.from_json.call(this, json);
    };

    return PresentationContextItemAc;

  })(PresentationContextItem);

  AbstractSyntaxItem = (function(_super) {
    __extends(AbstractSyntaxItem, _super);

    function AbstractSyntaxItem() {
      return AbstractSyntaxItem.__super__.constructor.apply(this, arguments);
    }

    AbstractSyntaxItem.prototype.type = 0x30;

    AbstractSyntaxItem.prototype.name = 'abstract_syntax';

    AbstractSyntaxItem.prototype._single_value = true;

    AbstractSyntaxItem.prototype.decode = function() {
      return this.value = this.ui_str(4);
    };

    AbstractSyntaxItem.prototype.encode = function() {
      return this.encode_value_str();
    };

    return AbstractSyntaxItem;

  })(Item);

  TransferSyntaxItem = (function(_super) {
    __extends(TransferSyntaxItem, _super);

    function TransferSyntaxItem() {
      return TransferSyntaxItem.__super__.constructor.apply(this, arguments);
    }

    TransferSyntaxItem.prototype.type = 0x40;

    TransferSyntaxItem.prototype.name = 'transfer_syntax';

    TransferSyntaxItem.prototype._single_value = true;

    TransferSyntaxItem.prototype.decode = function() {
      return this.value = this.ui_str(4);
    };

    TransferSyntaxItem.prototype.encode = function() {
      return this.encode_value_str();
    };

    return TransferSyntaxItem;

  })(Item);

  UserInformationItem = (function(_super) {
    __extends(UserInformationItem, _super);

    function UserInformationItem() {
      return UserInformationItem.__super__.constructor.apply(this, arguments);
    }

    UserInformationItem.prototype.type = 0x50;

    UserInformationItem.prototype.name = 'user_information';

    UserInformationItem.prototype.var_item_counts = {
      maximum_length: 1,
      asynchronous_operations_window: 1,
      implementation_class_uid: 1,
      implementation_version_name: 1,
      scp_scu_role_selection: -1
    };

    UserInformationItem.prototype.decode = function() {
      return this.decode_var_items(this._start + 4, this._end);
    };

    UserInformationItem.prototype.encode = function() {
      var _buffers, _header, _len, _rs;
      _buffers = ['', this.maximum_length.encode(), this.implementation_class_uid.encode(), this.asynchronous_operations_window.encode()];
      if (this.scp_scu_role_selection) {
        _buffers.push(Buffer.concat((function() {
          var _i, _len, _ref, _results;
          _ref = this.scp_scu_role_selection;
          _results = [];
          for (_i = 0, _len = _ref.length; _i < _len; _i++) {
            _rs = _ref[_i];
            _results.push(_rs.encode());
          }
          return _results;
        }).call(this)));
      }
      _buffers.push(this.implementation_version_name.encode());
      _len = total_length(_buffers);
      _header = Buffer.concat([new Buffer([this.type, 0]), mk_uint16(_len)]);
      _buffers[0] = _header;
      return Buffer.concat(_buffers);
    };

    return UserInformationItem;

  })(Item);

  MaximumLengthItem = (function(_super) {
    __extends(MaximumLengthItem, _super);

    function MaximumLengthItem() {
      return MaximumLengthItem.__super__.constructor.apply(this, arguments);
    }

    MaximumLengthItem.prototype.type = 0x51;

    MaximumLengthItem.prototype.name = 'maximum_length';

    MaximumLengthItem.prototype._single_value = true;

    MaximumLengthItem.prototype.decode = function() {
      return this.value = this._buff.readUInt32BE(this._start + 4);
    };

    MaximumLengthItem.prototype.encode = function() {
      var _vbuff;
      _vbuff = new Buffer(4);
      _vbuff.writeUInt32BE(this.value, 0);
      return Buffer.concat([new Buffer([this.type, 0]), mk_uint16(4), _vbuff]);
    };

    return MaximumLengthItem;

  })(Item);

  ImplementationClassUidItem = (function(_super) {
    __extends(ImplementationClassUidItem, _super);

    function ImplementationClassUidItem() {
      return ImplementationClassUidItem.__super__.constructor.apply(this, arguments);
    }

    ImplementationClassUidItem.prototype.type = 0x52;

    ImplementationClassUidItem.prototype.name = 'implementation_class_uid';

    ImplementationClassUidItem.prototype._single_value = true;

    ImplementationClassUidItem.prototype.decode = function() {
      return this.value = this.ui_str(4);
    };

    ImplementationClassUidItem.prototype.encode = function() {
      return this.encode_value_str();
    };

    return ImplementationClassUidItem;

  })(Item);

  AsynchronousOperationsWindowItem = (function(_super) {
    __extends(AsynchronousOperationsWindowItem, _super);

    function AsynchronousOperationsWindowItem() {
      return AsynchronousOperationsWindowItem.__super__.constructor.apply(this, arguments);
    }

    AsynchronousOperationsWindowItem.prototype.type = 0x53;

    AsynchronousOperationsWindowItem.prototype.name = 'asynchronous_operations_window';

    AsynchronousOperationsWindowItem.prototype.decode = function() {
      this.maximum_number_operations_invoked = this._buff.readUInt16BE(this._start + 4);
      return this.maximum_number_operations_performed = this._buff.readUInt16BE(this._start + 6);
    };

    AsynchronousOperationsWindowItem.prototype.to_json = function() {
      var _json;
      _json = AsynchronousOperationsWindowItem.__super__.to_json.call(this);
      _json.maximum_number_operations_invoked = this.maximum_number_operations_invoked;
      _json.maximum_number_operations_performed = this.maximum_number_operations_performed;
      return _json;
    };

    AsynchronousOperationsWindowItem.prototype.from_json = function(json) {
      this.maximum_number_operations_invoked = json.maximum_number_operations_invoked;
      this.maximum_number_operations_performed = json.maximum_number_operations_performed;
      return AsynchronousOperationsWindowItem.__super__.from_json.call(this, json);
    };

    AsynchronousOperationsWindowItem.prototype.encode = function() {
      var _vbuff;
      _vbuff = new Buffer(4);
      _vbuff.writeUInt16BE(this.maximum_number_operations_invoked, 0);
      _vbuff.writeUInt16BE(this.maximum_number_operations_performed, 2);
      return Buffer.concat([new Buffer([this.type, 0]), mk_uint16(4), _vbuff]);
    };

    return AsynchronousOperationsWindowItem;

  })(Item);

  ScpScuRoleSelectionItem = (function(_super) {
    __extends(ScpScuRoleSelectionItem, _super);

    function ScpScuRoleSelectionItem() {
      return ScpScuRoleSelectionItem.__super__.constructor.apply(this, arguments);
    }

    ScpScuRoleSelectionItem.prototype.type = 0x54;

    ScpScuRoleSelectionItem.prototype.name = 'scp_scu_role_selection';

    ScpScuRoleSelectionItem.prototype.decode = function() {
      var _end, _start, _uid_length;
      _uid_length = this._buff.readUInt16BE(this._start + 4);
      _start = this._start + 6;
      _end = this._start + 6 + _uid_length;
      this.sop_class_uid = trim_ui(this._buff.toString('binary', _start, _end));
      this.scu_role = this._buff[_end];
      return this.scp_role = this._buff[_end + 1];
    };

    ScpScuRoleSelectionItem.prototype.to_json = function() {
      var _json;
      _json = ScpScuRoleSelectionItem.__super__.to_json.call(this);
      _json.sop_class_uid = this.sop_class_uid;
      _json.scu_role = this.scu_role;
      _json.scp_role = this.scp_role;
      return _json;
    };

    ScpScuRoleSelectionItem.prototype.from_json = function(json) {
      this.sop_class_uid = json.sop_class_uid;
      this.scu_role = json.scu_role;
      this.scp_reole = json.scp_role;
      return ScpScuRoleSelectionItem.__super__.from_json.call(this, json);
    };

    ScpScuRoleSelectionItem.prototype.encode = function() {
      var _buffers, _header, _len;
      _buffers = ['', mk_uint16(this.sop_class_uid.length), new Buffer(this.sop_class_uid, 'binary'), new Buffer([this.scu_role, this.scp_role])];
      _len = total_length(_buffers);
      _header = Buffer.concat([new Buffer([this.type, 0]), mk_uint16(_len)]);
      _buffers[0] = _header;
      return Buffer.concat(_buffers);
    };

    return ScpScuRoleSelectionItem;

  })(Item);

  ImplementationVersionNameItem = (function(_super) {
    __extends(ImplementationVersionNameItem, _super);

    function ImplementationVersionNameItem() {
      return ImplementationVersionNameItem.__super__.constructor.apply(this, arguments);
    }

    ImplementationVersionNameItem.prototype.type = 0x55;

    ImplementationVersionNameItem.prototype.name = 'implementation_version_name';

    ImplementationVersionNameItem.prototype._single_value = true;

    ImplementationVersionNameItem.prototype.decode = function() {
      return this.value = this.ui_str(4);
    };

    ImplementationVersionNameItem.prototype.encode = function() {
      return this.encode_value_str();
    };

    return ImplementationVersionNameItem;

  })(Item);

  trim_ui = function(str) {
    var _len;
    _len = str.length;
    if (_len > 0 && str[_len - 1] === '\x00') {
      return str.slice(0, -1);
    } else {
      return str;
    }
  };

  PDU_BY_TYPE = {
    '01': PDUAssociateRq,
    '02': PDUAssociateAc
  };

  PDU_BY_NAME = {
    'association_rq': PDUAssociateRq,
    'association_ac': PDUAssociateAc
  };

  pdu_constructor = function(type) {
    var _hex;
    _hex = printf("%02X", type);
    return PDU_BY_TYPE[_hex];
  };

  pdu_by_name = function(arg) {
    if (arg instanceof PDU) {
      return arg;
    }
    return new PDU_BY_NAME[arg.name](arg);
  };

  ITEM_BY_TYPE = {
    '10': ApplicationContextItem,
    '20': PresentationContextItem,
    '21': PresentationContextItemAc,
    '30': AbstractSyntaxItem,
    '40': TransferSyntaxItem,
    '50': UserInformationItem,
    '51': MaximumLengthItem,
    '52': ImplementationClassUidItem,
    '53': AsynchronousOperationsWindowItem,
    '54': ScpScuRoleSelectionItem,
    '55': ImplementationVersionNameItem
  };

  ITEM_BY_NAME = {
    'application_context': ApplicationContextItem,
    'presentation_context': PresentationContextItem,
    'presentation_context_ac': PresentationContextItemAc,
    'abstract_syntax': AbstractSyntaxItem,
    'transfer_syntax': TransferSyntaxItem,
    'user_information': UserInformationItem,
    'maximum_length': MaximumLengthItem,
    'implementation_class_uid': ImplementationClassUidItem,
    'asynchronous_operations_window': AsynchronousOperationsWindowItem,
    'scp_scu_role_selection': ScpScuRoleSelectionItem,
    'implementation_version_name': ImplementationVersionNameItem
  };

  item_constructor = function(type) {
    var _hex;
    _hex = printf("%02X", type);
    return ITEM_BY_TYPE[_hex];
  };

  ZERO_BUFF = new Buffer(128);

  PDUEncoder = (function(_super) {
    __extends(PDUEncoder, _super);

    function PDUEncoder(options) {
      if (!(this instanceof PDUEncoder)) {
        return new PDUEncoder(options);
      }
      PDUEncoder.__super__.constructor.call(this, options);
      this._writableState.objectMode = true;
      this._readableState.objectMode = false;
    }

    PDUEncoder.prototype._transform = function(pdu, _, cb) {
      var err, __buff;
      try {
        __buff = pdu_by_name(pdu).encode();
        log.trace({
          length: __buff.length
        }, "_transform: emitting pdu buffer");
        this.push(__buff);
        return cb();
      } catch (_error) {
        err = _error;
        log.error({
          error: err
        }, "_transform: error");
        return cb(err);
      }
    };

    PDUEncoder.prototype._flush = function() {
      return log.debug("_flush");
    };

    return PDUEncoder;

  })(stream.Transform);

  mk_uint16 = function(num) {
    var _buff;
    _buff = new Buffer(2);
    _buff.writeUInt16BE(num, 0);
    return _buff;
  };

  mk_uint32 = function(num) {
    var _buff;
    _buff = new Buffer(4);
    _buff.writeUInt32BE(num, 0);
    return _buff;
  };

  exports.PDUDecoder = PDUDecoder;

  exports.PDUEncoder = PDUEncoder;

  exports.PDUAssociateRq = PDUAssociateRq;

  net = require('net');

  echo_scu = function(opts, cb) {
    var _aet, _conn, _dec, _enc, _local;
    _aet = opts.aet;
    _local = opts.local_aet;
    _conn = net.connect(opts, function() {
      console.log("connected to", opts);
      return _enc.write({
        "name": "association_rq",
        "called_aet_title": _aet,
        "calling_aet_title": _local,
        "application_context": "1.2.840.10008.3.1.1.1",
        "presentation_context": [
          {
            "id": 1,
            "abstract_syntax": "1.2.840.10008.1.1",
            "transfer_syntax": ["1.2.840.10008.1.2"]
          }
        ],
        "user_information": {
          "maximum_length": 16384,
          "implementation_class_uid": "1.2.40.0.13.1.1.47",
          "asynchronous_operations_window": {
            "maximum_number_operations_invoked": 0,
            "maximum_number_operations_performed": 0
          },
          "implementation_version_name": "node-dicom"
        }
      });
    });
    _enc = new PDUEncoder();
    _dec = new PDUDecoder();
    _enc.pipe(_conn);
    _conn.pipe(_dec);
    _conn.on('error', cb);
    _enc.on('error', cb);
    _dec.on('error', cb);
    return _dec.on('data', function(data) {
      return console.log("RECEIVED:", JSON.stringify(data, null, 2));
    });
  };

  if (require.main === module) {
    if (true) {
      echo_scu({
        host: "192.168.2.19",
        port: 11112,
        aet: "TESTME",
        local_aet: "XXX"
      }, function(err, data) {
        if (err) {
          console.log("ERROR: ", err);
          console.log("stack:", err.stack);
          process.exit(1);
        }
        return console.log("DATA:", data);
      });
    }
    if (false) {
      require('fs').createReadStream("/tmp/x.x").pipe(new PDUDecoder());
    }
  }

}).call(this);
