// dicom/lib/json/encoder.js
// Generated by CoffeeScript 1.8.0
(function() {
  var JsonEncoder, log, printf, stream, tags,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  stream = require("stream");

  printf = require("printf");

  tags = require("../../lib/tags");

  log = require("../logger")("json", "encoder");

  JsonEncoder = (function(_super) {
    __extends(JsonEncoder, _super);

    function JsonEncoder(options) {
      if (!(this instanceof JsonEncoder)) {
        return new JsonEncoder(options);
      }
      JsonEncoder.__super__.constructor.call(this, options);
      this._bulkdata_uri = options != null ? options.bulkdata_uri : void 0;
      this._writableState.objectMode = true;
      this._readableState.objectMode = false;
      this.depth = 0;
      this.fresh = true;
      this.ignore = 0;
    }

    JsonEncoder.prototype._transform = function(event, encoding, cb) {
      var command, err, _ref;
      try {
        if (log.debug()) {
          log.debug({
            command: event.command,
            element: (_ref = event.element) != null ? _ref.name : void 0,
            depth: this.depth
          }, "Json:_transform received dicom event");
        }
        command = event.command;
        switch (command) {
          case 'element':
            this.handle_element(event);
            break;
          case 'start_sequence':
            this.start_sequence(event);
            break;
          case 'end_sequence':
            this.end_sequence(event);
            break;
          case 'start_item':
            this.start_item(event);
            break;
          case 'end_item':
            this.end_item(event);
            break;
          case 'start_element':
            this.start_element(event);
            break;
          case 'end_element':
            this.end_element(event);
            break;
          default:
            if (log.trace()) {
              log.trace({
                command: command
              }, "_transform: ignoring");
            }
        }
        return cb(null);
      } catch (_error) {
        err = _error;
        log.error(err);
        return cb(err);
      }
    };

    JsonEncoder.prototype._flush = function(cb) {
      this.push("}\n");
      return cb(null);
    };

    JsonEncoder.prototype.handle_element = function(event) {
      var key, obj, start;
      if (this.ignore) {
        return;
      }
      key = printf('"%08X"', event.element.tag);
      key = printf("%*s", key, key.length + this.depth);
      obj = {
        vr: event.vr.name
      };
      if (event.vr.base64_values) {
        obj.InlineBinary = event.vr.values();
      } else {
        obj.Value = event.vr.values();
      }
      start = ',\n';
      if (this.fresh) {
        start = '{\n';
        this.fresh = false;
      }
      return this.push(printf('%s%s: %s', start, key, JSON.stringify(obj)));
    };

    JsonEncoder.prototype.start_sequence = function(event) {
      var key, start;
      if (this.ignore) {
        return;
      }
      key = printf('"%08X"', event.element.tag);
      key = printf("%*s", key, key.length + this.depth);
      start = ',\n';
      if (this.fresh) {
        start = '{\n';
      }
      this.push(printf('%s%s: {"vr":"SQ", "Value": [', start, key));
      this.fresh = true;
      return this.depth++;
    };

    JsonEncoder.prototype.end_sequence = function(event) {
      if (this.ignore) {
        return;
      }
      this.fresh = false;
      this.push(']}');
      return this.depth--;
    };

    JsonEncoder.prototype.start_item = function(event) {
      var bd_uri;
      if (this.ignore) {
        return;
      }
      if (!this.fresh) {
        this.push(",");
      }
      this.fresh = true;
      if (event.bulkdata_offset && event.bulkdata_length) {
        bd_uri = this._bulkdata_uri + "?offset=" + event.bulkdata_offset + "&length=" + event.bulkdata_length;
        this.push(printf('{"BulkDataURI":%s', JSON.stringify(bd_uri)));
        return this.fresh = false;
      }
    };

    JsonEncoder.prototype.end_item = function(event) {
      if (this.ignore) {
        return;
      }
      if (this.fresh) {
        this.push("{}");
      } else {
        this.push("}");
      }
      return this.fresh = false;
    };

    JsonEncoder.prototype.start_element = function(event) {
      var bd_uri, key, start;
      if (this._bulkdata_uri) {
        key = printf('"%08X"', event.element.tag);
        key = printf("%*s", key, key.length + this.depth);
        start = ',\n';
        if (this.fresh) {
          start = '{\n';
        }
        if (event.bulkdata_offset && event.bulkdata_length) {
          bd_uri = this._bulkdata_uri + "?offset=" + event.bulkdata_offset + "&length=" + event.bulkdata_length;
          this.push(printf('%s%s: {"vr":"%s","BulkDataURI":%s', start, key, event.vr.name, JSON.stringify(bd_uri)));
        } else {
          this.push(printf('%s%s: {"vr":"%s","DataFragment": [', start, key, event.vr.name));
        }
        this.fresh = true;
        return this.depth++;
      } else {
        return this.ignore++;
      }
    };

    JsonEncoder.prototype.end_element = function(event) {
      if (this.ignore) {
        this.ignore--;
        return;
      }
      if (this._bulkdata_uri) {
        this.fresh = false;
        if (event.bulkdata_offset && event.bulkdata_length) {
          this.push('}');
        } else {
          this.push(']}');
        }
        return this.depth--;
      }
    };

    return JsonEncoder;

  })(stream.Transform);

  module.exports = JsonEncoder;

}).call(this);
