// dicom/lib/json/source.js
// Generated by CoffeeScript 1.8.0
(function() {
  var EmitStack, ItemEntry, JsonSource, SeqEntry, data, log, simple_data, source, stream, tags, uids, util, vrs,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  stream = require("stream");

  util = require("util");

  tags = require("../../lib/tags");

  uids = require("../../lib/uids");

  vrs = require("../../lib/vrs");

  log = require("../logger")("json", "source");

  ItemEntry = (function() {
    function ItemEntry(data, _end_event) {
      var k, v;
      this._end_event = _end_event;
      this._data = {};
      for (k in data) {
        v = data[k];
        this._data[tags.for_tag(k).tag_str] = v;
      }
      this._queue = (function() {
        var _results;
        _results = [];
        for (k in this._data) {
          _results.push(k);
        }
        return _results;
      }).call(this);
      this._queue.sort();
      this._queue.reverse();
    }

    ItemEntry.prototype.unshift = function() {
      return this._queue.pop();
    };

    ItemEntry.prototype.end_event = function() {
      return this._end_event;
    };

    ItemEntry.prototype.data = function(k) {
      var el, v, _ref;
      v = this._data[k];
      el = tags.for_tag(k);
      if (util.isArray(v)) {
        v = {
          vr: el.vr,
          Value: v
        };
      } else if ((_ref = typeof v) === 'string' || _ref === 'number') {
        v = {
          vr: el.vr,
          Value: [v]
        };
      } else if (typeof v === 'object') {
        if (v.BulkDataURI) {
          throw new vrs.DicomError("can not emit json model with bulkdata uri: " + v);
        }
        if (v.vr == null) {
          v.vr = el.vr;
        }
      } else {
        throw new vrs.DicomError("can not recognize dicom json model: " + v);
      }
      return [el, v];
    };

    return ItemEntry;

  })();

  SeqEntry = (function() {
    function SeqEntry(items) {
      var x;
      this._queue = (function() {
        var _i, _len, _results;
        _results = [];
        for (_i = 0, _len = items.length; _i < _len; _i++) {
          x = items[_i];
          _results.push(x);
        }
        return _results;
      })();
      this._queue.reverse();
      log.trace({
        length: this._queue.length
      }, "SeqEntry");
    }

    SeqEntry.prototype.unshift = function() {
      return this._queue.pop();
    };

    SeqEntry.prototype.end_event = function() {
      return new vrs.DicomEvent(tags.SequenceDelimitationItem, null, null, "end_sequence");
    };

    return SeqEntry;

  })();

  EmitStack = (function() {
    function EmitStack() {
      this._len_1 = -1;
      this._stack = [];
    }

    EmitStack.prototype.push = function(data, end_event) {
      this._len_1++;
      return this._stack.push(new ItemEntry(data, end_event));
    };

    EmitStack.prototype.push_seq = function(items) {
      this._len_1++;
      return this._stack.push(new SeqEntry(items));
    };

    EmitStack.prototype.pop = function() {
      this._len_1--;
      return this._stack.pop();
    };

    EmitStack.prototype.top = function() {
      return this._stack[this._len_1];
    };

    EmitStack.prototype.unshift = function() {
      return this.top().unshift();
    };

    EmitStack.prototype.data = function(k) {
      return this.top().data(k);
    };

    EmitStack.prototype.eof = function() {
      return this._len_1 === -1;
    };

    return EmitStack;

  })();

  JsonSource = (function(_super) {
    __extends(JsonSource, _super);

    function JsonSource(data, options) {
      var ts, ts_name;
      if (!(this instanceof JsonSource)) {
        return new JsonSource(data, options);
      }
      if (options == null) {
        options = {};
      }
      options.objectMode = true;
      JsonSource.__super__.constructor.call(this, options);
      this._stack = new EmitStack();
      this._stack.push(data, null);
      ts_name = options != null ? options.transfer_syntax : void 0;
      if (!ts_name) {
        ts_name = 'ExplicitVRLittleEndian';
      }
      ts = uids.for_uid(ts_name);
      this._context = new vrs.Context({}, ts.make_context());
      log.trace({
        context: this._context
      }, "JsonSource context");
    }

    JsonSource.prototype._read = function(size) {
      var el, entry, err, k, obj, read_more, v, _ref;
      try {
        log.trace({
          size: size
        }, "JsonSource _read");
        read_more = true;
        while (read_more) {
          if (this._stack.eof()) {
            log.trace("_stack eof: we are done");
            this.push(null);
            read_more = false;
            return;
          } else {
            k = this._stack.unshift();
            if (k != null) {
              if (typeof k === 'string') {
                _ref = this._stack.data(k), el = _ref[0], v = _ref[1];
                obj = this._dicom_event(el, v);
                log.trace(typeof obj.log_summary === "function" ? obj.log_summary() : void 0, "emitting");
                read_more = this.push(obj);
                if (obj.command === 'start_sequence') {
                  log.trace(v, "pushing sequence items");
                  this._stack.push_seq(v.Value);
                }
              } else {
                this._stack.push(k, new vrs.DicomEvent(tags.ItemDelimitationItem, null, null, "end_item"));
                obj = new vrs.DicomEvent(tags.Item, null, null, "start_item");
                log.trace(typeof obj.log_summary === "function" ? obj.log_summary() : void 0, "emitting start item");
                read_more = this.push(obj);
              }
            } else {
              entry = this._stack.pop();
              obj = entry.end_event();
              if (obj) {
                log.trace(typeof obj.log_summary === "function" ? obj.log_summary() : void 0, "emitstack end event");
                read_more = this.push(obj);
              }
            }
          }
        }
        return void 0;
      } catch (_error) {
        err = _error;
        return this.emit('error', err);
      }
    };

    JsonSource.prototype._dicom_event = function(el, v) {
      if (this._is_seq_value(v)) {
        if (v.vr === 'UN' || v.vr === 'SQ') {
          return new vrs.DicomEvent(el, vrs.for_name(v.vr, this._context), null, "start_sequence");
        } else {
          throw new vrs.DicomError("can not emit sequence values for vr " + v.vr + " tag " + el.tag_str);
        }
      } else {
        return new vrs.DicomEvent(el, vrs.for_name(v.vr, this._context, null, v.Value), null, "element");
      }
    };

    JsonSource.prototype._is_seq_value = function(v) {
      var is_seq, _i, _len, _ref, _v;
      is_seq = true;
      _ref = v.Value;
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        _v = _ref[_i];
        if (!((typeof _v === 'object') && !util.isArray(_v))) {
          is_seq = false;
        }
      }
      return is_seq;
    };

    return JsonSource;

  })(stream.Readable);

  module.exports = JsonSource;

  if (require.main === module) {
    tags = require("../../lib/tags");
    data = {
      "00100020": {
        vr: "LO",
        Value: ["007"]
      },
      "00100021": {
        vr: "LO",
        Value: ["MI6"]
      },
      "00101002": {
        vr: "SQ",
        Value: [
          {
            "00100020": {
              vr: "LO",
              Value: ["0815"]
            },
            "00100021": {
              vr: "LO",
              Value: ["BND"]
            }
          }
        ]
      }
    };
    simple_data = {
      "PatientID": "007",
      "IssuerOfPatientID": "MI6",
      "OtherPatientIDsSequence": [
        {
          "PatientID": "0815",
          "IssuerOfPatientID": "BND"
        }
      ]
    };
    source = new JsonSource(data);
    source.pipe(process.stdout);
  }

}).call(this);
